src/Compilers/CSharp/Portable/Binder/Semantics/Operators/UnaryOperatorOverloadResolution.cs

```diff
@@ -78,31 +78,55 @@ public void UnaryOperatorOverloadResolution(UnaryOperatorKind kind, BoundExpress
             // SPEC: better than all other function members, then the function member invocation is ambiguous and a binding-time 
             // SPEC: error occurs.
 
-            // UNDONE: This is a naive quadratic algorithm; there is a linear algorithm that works. Consider using it.
+            int i;
             var candidates = result.Results;
-            for (int i = 1; i < candidates.Count; ++i)
+            for (i = 0; i < candidates.Count; i++)
             {
-                if (candidates[i].Kind != OperatorAnalysisResultKind.Applicable)
+                if (candidates[i].Kind == OperatorAnalysisResultKind.Applicable)
+                {
+                    break;
+                }
+            }
+
+            for (int j = i + 1; j < candidates.Count; j++)
+            {
+                if (candidates[j].Kind == OperatorAnalysisResultKind.Inapplicable)
+                {
+                    continue;
+                }
+
+                var better = BetterOperator(candidates[i].Signature, candidates[j].Signature, operand, ref useSiteDiagnostics);
```

Comment:
`operand` here, versus `left` and `right` in binary case. That's the only difference between the two implementations. I wonder if they could be factored together.